src/time_integration/paired_explicit_runge_kutta/paired_explicit_runge_kutta.jl

```diff
@@ -108,21 +108,48 @@ function solve!(integrator::AbstractPairedExplicitRKIntegrator)
 
     @trixi_timeit timer() "main loop" while !integrator.finalstep
         step!(integrator)
-    end # "main loop" timer
+    end
 
     return TimeIntegratorSolution((first(prob.tspan), integrator.t),
                                   (prob.u0, integrator.u),
                                   integrator.sol.prob)
 end
 
+# Function that computes the first stage of a general PERK method
+@inline function PERK_k1!(integrator::AbstractPairedExplicitRKIntegrator, p)
+    integrator.f(integrator.k1, integrator.u, p, integrator.t)
+end
+
+@inline function PERK_k2!(integrator::AbstractPairedExplicitRKSingleIntegrator, p, c)
+    @threaded for i in eachindex(integrator.du)
+        integrator.u_tmp[i] = integrator.u[i] + c[2] * integrator.dt * integrator.k1[i]
+    end
+
+    integrator.f(integrator.du, integrator.u_tmp, p,
+                 integrator.t + c[2] * integrator.dt)
+end
+
+@inline function PERK_ki!(integrator::AbstractPairedExplicitRKSingleIntegrator, p, c,
+                          a_matrix, stage)
+    # Construct current state
+    @threaded for i in eachindex(integrator.u)
+        integrator.u_tmp[i] = integrator.u[i] +
+                              integrator.dt *
+                              (a_matrix[stage - 2, 1] * integrator.k1[i] +
+                               a_matrix[stage - 2, 2] * integrator.du[i])
+    end
+
+    integrator.f(integrator.du, integrator.u_tmp, p,
+                 integrator.t + c[stage] * integrator.dt)
```

Comment:
This does look indeed more compact. If I understand this correctly, we should still be able to access the k value of a specific stage by just calculating `integrator.du * integrator.dt` after this line, right?

Reply:
Yes exactly, so we spare the access to `du` by multiplying `dt` repeatedly. As memory accesses are usually way more expensive than an additional floating point multiplication, this should be more efficient.

Reply:
Got it!